providers/keycloak.go

```diff
@@ -11,7 +11,7 @@ import (
 
 type KeycloakProvider struct {
 	*ProviderData
-	Group string
+	Groups []string
```

Comment:
This would be better as `Groups map[string]struct{}`. This way you can use the built in map lookup instead of having to write the contains function

In `SetGroups` you would do

```
p.Groups = make(map[string}struct{})
for _, group := range groups {
  p.Groups[group] = struct{}{}
}
```

And then instead of contains you would just do a look up

```
if _, ok := p.Groups[group]; ok {
  found = true
  break
}

Reply:
If we use this central method: https://github.com/oauth2-proxy/oauth2-proxy/blob/3ff0c23a9e418875cdd7460b0a6a5b40dea71770/providers/provider_data.go#L54

Then all of this logic is handled by the central `Authorization` implementation. `Keycloak` is technically already allowing multiple groups now via `allowed-group` config option via that route I think?

To make the other option backwards compatible with it, we should do something like this to make legacy flags work: https://github.com/oauth2-proxy/oauth2-proxy/blob/3ff0c23a9e418875cdd7460b0a6a5b40dea71770/pkg/validation/options.go#L261

Then I think all the groups tracking & contains methods on the Keycloak provider can go away.

I think the big area this PR needs to adapt to the new paradigms is in `GetEmailAddress` (which should switch to `EnrichSessionState`), is adding the groups returned from `ValidateURL` into the `session.Groups` list.